### ISSUE TITLE
__init__ literal type hint gets converted to str in second method

### ISSUE THREAD

**Issue by thomasaarholt** [microsoft/pyright#4541]
**Describe the bug**
In `Bar.__init__` below, `self.foo` is implicitly typed by the type hint in its definition. When I pass `self.foo` to a function inside `__init__`, type checking passes. When I pass it in a second method, it seems to not know that is typed as a literal, but does think it is a `str`, raising an error. 

If I explicitly type hint`self.foo: FOO_TYPE`, it works as expected.

Thinking about this - is this because pyright can't guarantee that `self.foo` hasn't been mutated before it is called in `Bar.method`?

**To Reproduce**
```python
from typing import Literal, TypeAlias

FOO_TYPE: TypeAlias = Literal["a", "b", "c"]


def func(foo: FOO_TYPE):
    ...

class Bar:
    def __init__(self, foo: FOO_TYPE):
        self.foo = foo
        reveal_type(self.foo) # Type of "self.foo" is "Literal['a', 'b', 'c']"
        func(foo=self.foo) # Works here

    def method(self):
        reveal_type(self.foo) # Type of "self.foo" is "str"
        func(foo=self.foo) # Fails here
#                ^^^^^^^^
# Argument of type "str" cannot be assigned to parameter "foo" of type 
# "FOO_TYPE" in function "func"
#   Type "str" cannot be assigned to type "FOO_TYPE"
#     "str" cannot be assigned to type "Literal['a']"
#     "str" cannot be assigned to type "Literal['b']"
#     "str" cannot be assigned to type "Literal['c']"

# Second case, added explicit type hint, now works
class Baz:
    def __init__(self, foo: FOO_TYPE):
        self.foo: FOO_TYPE = foo # Added type hint, now no error

    def method(self):
        self.foo
        func(foo=self.foo)
```

**VS Code extension or command-line**
Running VSCode insiders, Python Extension v2023.1.41

**Comment by erictraut** [comment]
Yes, this behavior is correct. The type of instance variable `foo` is not declared, so its type is inferred based on assignments within the class. It is assigned a value in the `__init__` method. Its inferred type is `str`. However, within the `__init__` method, its type is locally narrowed to type `Literal["a", "b", "c"]` due to type narrowing on assignment.

If your intent is for the instance variable `foo` to be limited to literal values, you must declare it as such:
```python
self.foo: FOO_TYPE = foo
```

For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#literals) and [this documentation](https://github.com/microsoft/pyright/blob/main/docs/mypy-comparison.md#assignment-based-narrowing-for-literals).

**Comment by thomasaarholt** [comment]
That documentation was excellent! I'm extremely impressed with the thoroughness and reason of your replies. Thank you.
